adapter-base/src/main/java/org/eclipse/hono/adapter/AbstractProtocolAdapterBase.java

```diff
@@ -936,8 +936,10 @@ public void registerLivenessChecks(final HealthCheckHandler handler) {
      */
     protected Future<Void> sendConnectedEvent(final String remoteId, final Device authenticatedDevice, final SpanContext context) {
         if (this.connectionEventProducer != null) {
-            return getTenantClient().get(authenticatedDevice.getTenantId(), context)
-                    .map(this::getEventSender)
+            return Optional.ofNullable(authenticatedDevice)
+                    .map(device -> getTenantClient().get(device.getTenantId(), context)
+                            .map(this::getEventSender))
+                    .orElseGet(() -> Future.succeededFuture(null))
```

Comment:
returning `null` here violates the *ConnectionEventProducer.Context.getMessageSenderClient* contract.
I wonder if it makes any sense to invoke this method without an authenticated device? Otherwise, we could also allow this method to be invoked with any kind of Device, either authenticated or unauthenticated ...

Reply:
> returning `null` here violates the *ConnectionEventProducer.Context.getMessageSenderClient* contract.

I've changed that contract in this PR to allow `null` to be returned.
I can add a note in the release notes to make this more obvious.

> I wonder if it makes any sense to invoke this method without an authenticated device? Otherwise, we could also allow this method to be invoked with any kind of Device, either authenticated or unauthenticated ...

The original method here worked fine with unauthenticated devices:
```
    /**
     * Trigger the creation of a "connected" event.
     * 
     * @param remoteId The remote ID.
     * @param authenticatedDevice The (optional) authenticated device.
     * 
     * @see ConnectionEventProducer
     * @see ConnectionEventProducer#connected(String, String, Device, JsonObject)
     */
    protected Future<?> sendConnectedEvent(final String remoteId, final Device authenticatedDevice) {
        if (this.connectionEventProducer != null) {
            return this.connectionEventProducer.connected(remoteId, getTypeName(), authenticatedDevice, null);
        } else {
            return Future.succeededFuture();
        }
    }
```
FMPOV, it makes sense to allow unauthenticated devices here for the case that the `LoggingConnectionEventProducer` is used.

Reply:
> I've changed that contract in this PR to allow null to be returned.

I do not quite see the point of doing so. Don't we *always* have an event sender, i.e. either a sender for tenant specific messaging infrastructure or the fall back sender (AMQP)?

Reply:
> FMPOV, it makes sense to allow unauthenticated devices here for the case that the LoggingConnectionEventProducer is used.

Not sure if we are on the same page here. My point is to change the signature of *AbstractProtocolAdapterBase.sendConnectedEvent()* to no longer accept `null` for the *authenticatedDevice* param but instead rename the param to *device* and require it to be non-null. Clients of this method then need to be changed to either pass in the authenticated device or a `Device` instance created from the tenant-id and device-id specified in the target address.

Reply:
> Don't we *always* have an event sender, i.e. either a sender for tenant specific messaging infrastructure or the fall back sender (AMQP)?

FMPOV we always need a tenant object to be sure to get a valid event sender.
The `AbstractProtocolAdapterBase#getEventSender` method requires a non-null tenant object as parameter.
And an AMQP event sender can't be used as fallback if no AMQP messaging configuration exists (i.e. if only Kafka is used).

> My point is to change the signature of *AbstractProtocolAdapterBase.sendConnectedEvent()* to no longer accept `null` for the *authenticatedDevice* param [...].

The PR here is an alternative approach to that. It keeps the original support for unauthenticated devices in the `sendConnectedEvent()` method (useful for the `LoggingConnectionEventProducer` case) and doesn't require the protocol adapter code to be changed.

Reply:
my bad, I was assuming that we always know the tenant-id from the target address but we are talking about the *connected* event. So there is no target address involved here and when the device is not authenticated, then we also do not know the tenant ...